vpr/SRC/route/route_tree_timing.c

```diff
@@ -479,16 +562,600 @@ void free_route_tree(t_rt_node * rt_node) {
 }
 
 void update_net_delays_from_route_tree(float *net_delay,
-		t_rt_node ** rt_node_of_sink, int inet) {
+		const t_rt_node* const * rt_node_of_sink, int inet) {
 
 	/* Goes through all the sinks of this net and copies their delay values from *
 	 * the route_tree to the net_delay array.                                    */
 
-	unsigned int isink;
-	t_rt_node *sink_rt_node;
+	for (unsigned int isink = 1; isink < g_clbs_nlist.net[inet].pins.size(); isink++) {
+		net_delay[isink] = rt_node_of_sink[isink]->Tdel;
+	}
+}
+
+void update_remaining_net_delays_from_route_tree(float* net_delay, 
+		const t_rt_node* const * rt_node_of_sink, const vector<int>& remaining_sinks) {
+
+	/* Like update_net_delays_from_route_tree, but only updates the sinks that were not already routed              *
+	 * this function doesn't actually need to know about the net, just what sink pins need their net delays updated */
+
+	for (int sink_pin : remaining_sinks)
+		net_delay[sink_pin] = rt_node_of_sink[sink_pin]->Tdel;
+}
+
+
+/***************  Conversion between traceback and route tree *******************/
+t_rt_node* traceback_to_route_tree(int inet) {
+
+	/* Builds a skeleton route tree from a traceback                                      *
+	 * does not calculate R_upstream, C_downstream, or Tdel at all (left uninitialized)   *
+	 * returns the root of the converted route tree      								  *
+	 * initially points at the traceback equivalent of root 							  */
+
+	t_trace* head {trace_head[inet]};
+	// always called after the 1st iteration, so should exist
+	assert(head != nullptr);
+
+	t_rt_node* rt_root {init_route_tree_to_source(inet)};
+
+	t_rt_node* parent_node {rt_root};
+	// prev edge is always a dangling edge, waiting to point to next child, always belongs to parent_node
+	t_linked_rt_edge* prev_edge {alloc_linked_rt_edge()};	
+	prev_edge->iswitch = head->iswitch;
+	prev_edge->next = nullptr;
+	rt_root->u.child_list = prev_edge;
+
+	t_trace* tail;
+
+	while (true) {	// while the last sink has not been reached (tree has not been fully drawn)
```

Comment:
Would it be possible to refactor this so the exit condition check is in the while conditional? Same with loop below.